src/CodeGen_LLVM.cpp

```diff
@@ -1177,6 +1177,127 @@ void CodeGen_LLVM::optimize_module() {
+
+    PassBuilder::OptimizationLevel Level = PassBuilder::OptimizationLevel::O3;
+
+    // FIXME: Some bits in asan/tsan are different before 9.0
+    if (get_target().has_feature(Target::ASAN)) {
+        PB.registerPipelineStartEPCallback([&](ModulePassManager &MPM) {
+            MPM.addPass(
+                RequireAnalysisPass<ASanGlobalsMetadataAnalysis, llvm::Module>());
+        });
+        bool Recover = true;
+        bool UseAfterScope = true;
+        PB.registerOptimizerLastEPCallback(
+            [Recover, UseAfterScope](FunctionPassManager &FPM,
+                                     PassBuilder::OptimizationLevel Level) {
+                FPM.addPass(AddressSanitizerPass(
+                    /*CompileKernel=*/false, Recover, UseAfterScope));
+            });
+        bool ModuleUseAfterScope = false;
+        bool UseOdrIndicator = true;
+        PB.registerPipelineStartEPCallback(
+            [Recover, ModuleUseAfterScope,
+             UseOdrIndicator](ModulePassManager &MPM) {
+                MPM.addPass(ModuleAddressSanitizerPass(
+                    /*CompileKernel=*/false, Recover, ModuleUseAfterScope,
+                    UseOdrIndicator));
+            });
+    }
+
+    if (get_target().has_feature(Target::TSAN)) {
+        PB.registerOptimizerLastEPCallback(
+            [](FunctionPassManager &FPM, PassBuilder::OptimizationLevel Level) {
+                FPM.addPass(ThreadSanitizerPass());
+            });
+    }
+
+    for (llvm::Module::iterator i = module->begin(); i != module->end(); i++) {
+        if (get_target().has_feature(Target::ASAN)) {
+            i->addFnAttr(Attribute::SanitizeAddress);
+        }
+        if (get_target().has_feature(Target::TSAN)) {
+            // Do not annotate any of Halide's low-level synchronization code as it has
+            // tsan interface calls to mark its behavior and is much faster if
+            // it is not analyzed instruction by instruction.
+            if (!(i->getName().startswith("_ZN6Halide7Runtime8Internal15Synchronization") ||
+                  // TODO: this is a benign data race that re-initializes the detected features;
+                  // we should really fix it properly inside the implementation, rather than disabling
+                  // it here as a band-aid.
+                  i->getName().startswith("halide_default_can_use_target_features") ||
+                  i->getName().startswith("halide_mutex_") ||
+                  i->getName().startswith("halide_cond_"))) {
+                i->addFnAttr(Attribute::SanitizeThread);
+            }
+        }
+    }
+
+    MPM = PB.buildPerModuleDefaultPipeline(Level, debug_pass_manager);
+
+    // Verify?
+    MPM.addPass(VerifierPass());
+
+    MPM.run(*module, MAM);
+
+    // Verify?
+    if (llvm::verifyModule(*module, &errs()))
+      report_fatal_error("Transformation resulted in an invalid module\n");
+
+#else
+    // FIXME: Perhaps best to keep the legacy pass manager for older LLVM versions?
```

Comment:
I'm not clear on the status of the new pass manager in LLVM7 and LLVM8.

It is certain that the PTO flags are not there, because I only added them after LLVM8 was released. And I expect the code for adding sanitizers is different, because some pass names and APIs changed.
To ensure no regressions, I'd keep the use of the legacy pass manager for the older LLVM versions.